gpbft/gpbft.go

```diff
@@ -1418,36 +1417,35 @@ func newBroadcastState() *broadcastState {
 //   - https://github.com/filecoin-project/FIPs/discussions/809#discussioncomment-10409902
 //   - https://github.com/filecoin-project/FIPs/discussions/809#discussioncomment-10424988
 func (bs *broadcastState) record(mb *MessageBuilder) {
+	if bs.decide != nil {
+		return
+	}
 	switch mb.Payload.Phase {
+	case QUALITY_PHASE:
+		bs.quality = mb
 	case DECIDE_PHASE:
 		// Clear all previous messages, as only DECIDE message need to be rebroadcasted.
-		// Note that DECIDE message is not associated to any round, and is always
-		// broadcasted using round zero.
-		clear(bs.messagesByRound)
-		bs.messagesByRound[0] = []*MessageBuilder{mb}
-	case QUALITY_PHASE, CONVERGE_PHASE, PREPARE_PHASE, COMMIT_PHASE:
-		bs.messagesByRound[mb.Payload.Round] = append(bs.messagesByRound[mb.Payload.Round], mb)
-		// Remove all messages that are older than the latest two rounds, except QUALITY
-		// which only appears in round 0.
-		for round := int(mb.Payload.Round) - 2; round >= 0; round-- {
-			switch redundantRound := uint64(round); redundantRound {
-			case 0:
-				var found bool
-				for i, mb := range bs.messagesByRound[0] {
-					if mb.Payload.Phase == QUALITY_PHASE {
-						bs.messagesByRound[0] = bs.messagesByRound[0][i : i+1 : 1]
-						found = true
-						break
-					}
-				}
-				if !found {
-					log.Warn("No QUALITY message found for round 0 while trimming rebroadcast messages")
-					delete(bs.messagesByRound, redundantRound)
-				}
-			default:
-				delete(bs.messagesByRound, redundantRound)
-			}
+		// Note that DECIDE message is not associated to any round.
+		bs.currentRound = nil
+		bs.previousRound = nil
+		bs.quality = nil
+		bs.decide = mb
+	case CONVERGE_PHASE, PREPARE_PHASE, COMMIT_PHASE:
+		if mb.Payload.Round < bs.round {
+			log.Warnw("Unexpected broadcast of a message for a prior round")
```

Comment:
```suggestion
			log.Warnw("Unexpected broadcast of a message for a prior round", "latestRecordedRound", round, "messageRound", mb.Payload.Round)
```

Or use `log.Warn` if you think the suggested log line is an overkill.